### PR TITLE
feat: show unique HPs for Custom Searcher trials on experiment table [MLG-343]

### PR DESCRIPTION
**Pull request by tayritenour**
## Description
feat: show HPs for Custom Searcher trials on experiment table [MLG-343]

Before:
before

After:
after

## Test Plan

Run a custom searcher

```shell
cd examples/custom_search_method/asha_search_method
det experiment create remote_search_runner/searcher.yaml .
```

Before this change, the hyperparameters wouldn't display in the experiment table. Now they will.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket

[MLG-343]: https://determinedai.atlassian.net/browse/MLG-343?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by garrett361 — COMMENTED** [review]
I'm not sure the `filter out dictionary results that aren't leaf elements` commit is working. Or at least not functioning as I thought it would.  I rebuilt the web ui and used devcluster to test a custom searcher experiment with a subdictionary in `hyperparameters` and the dictionary is still printed out in the web ui.

Specifically in the asha custom searcher example I edited the `sample_params` function as in
```python
import copy
def sample_params() -> Dict[str, object]:
    hparams = {
        "global_batch_size": 64,
        "n_filters1": random.randint(8, 64),
        "n_filters2": random.randint(8, 72),
        "learning_rate": 10 ** random.uniform(-4.0, 0.0),
        "dropout1": random.uniform(0.2, 0.8),
        "dropout2": random.uniform(0.2, 0.8),
    }
    hparams["hparams"] = copy.deepcopy(hparams)
    return hparams

```
and expected to see (for instance) both `dropout` and `hparams.dropout` in as table columns, but not `hparams` itself as a column, and instead I see the below.
Scherm­afbeelding 2023-03-10 om 10 13 35

**Comment by tayritenour** [comment]
I added the ability to filter out by keys that are unique and it should no longer show the serialized dictionaries as you pointed out @garrett361
Screenshot 2023-03-10 at 2 29 35 PM

File: webui/react/src/pages/ExperimentDetails/ExperimentVisualization/LearningCurve.tsx [hunk 1]
```diff
@@ -74,12 +77,58 @@ const LearningCurve: React.FC<Props> = ({
   const hasTrials = trialHps.length !== 0;
   const isExperimentTerminal = terminalRunStates.has(experiment.state as RunState);
 
-  const hyperparameters = useMemo(() => {
-    return fullHParams.reduce((acc, key) => {
-      acc[key] = experiment.hyperparameters[key];
+  const get_custom_search_varying_hps = (trialHps: TrialHParams[]) => {
```
Comment (hkang1) [inline webui/react/src/pages/ExperimentDetails/ExperimentVisualization/LearningCurve.tsx:80]: **nit chore:** we use camel case for function names - `getCustomSearchVaryingHps`

File: webui/react/src/pages/ExperimentDetails/ExperimentVisualization/LearningCurve.tsx [hunk 2]
```diff
@@ -74,12 +77,58 @@ const LearningCurve: React.FC<Props> = ({
...
+     * definition of hyperparameters and determine what should be shown more dynamically.
+     *
+     * Note: If we support the other tabs in the future for Custom Searchers
+     * such as HpParallelCoordinates, HpScatterPlots, and HpHeatMaps, we will need to
+     * generalize this logic a bit.
+     */
+    const uniq = trialHps.reduce((acc, d) => {
+      Object.keys(d.hparams).forEach((key: RecordKey) => {
+        const value = d.hparams[key];
+        if (!isNumber(value)) {
```
Comment (garrett361) [inline webui/react/src/pages/ExperimentDetails/ExperimentVisualization/LearningCurve.tsx:95]: I think this check will be an issue in general, since there will be important non-numerical hps to display.  For DS in particular, there are boolean hps that may differ across trials, but more generally there may be important strings, lists, etc.

Testing by modifying  `examples/custom_search_method/asha_search_method/remote_search_runner/run_experiment.py`
to
```python
def sample_params() -> Dict[str, object]:
    subdict = {
        "float": random.uniform(0.0, 1.0),
        "int": random.randint(0, 16),
        "str": random.choice("this is a string".split()),
        "list": [random.randint(0, n) for n in range(3)],
        "bool": random.choice([True, False]),
        "dict": {n: random.randint(0, n) for n in range(3)},
    }

    hparams = {
        "global_batch_size": 64,
        "n_filters1": random.randint(8, 64),
        "n_filters2": random.randint(8, 72),
        "learning_rate": 10 ** random.uniform(-4.0, 0.0),
        "dropout1": random.uniform(0.2, 0.8),
        "dropout2": random.uniform(0.2, 0.8),
        "subdict": subdict,
    }
    return hparams


```

File: webui/react/src/pages/ExperimentDetails/ExperimentVisualization/LearningCurve.tsx (hunk 1 above)
Comment (hkang1) [inline webui/react/src/pages/ExperimentDetails/ExperimentVisualization/LearningCurve.tsx:80]: **suggestion:** this function looks like it is not dependent on anything inside the component `LearningCurve` and is purely reliant on the input parameter of `trialHps`. If that is the case, this function would be better to be defined just outside of the React component. It can be moved just above the start of `LearningCurve`.

**Review by hkang1 — COMMENTED** [review]
Thanks for making these updates to support custom seacher hps to show up! Left some comments and improvements below.

**Review by garrett361 — APPROVED** [review]
Did the same checks I noted in my previous comment and the table looks as I expected.  Great work!

File: webui/react/src/pages/ExperimentDetails/ExperimentVisualization/LearningCurve.tsx [hunk 3]
```diff
@@ -74,12 +77,58 @@ const LearningCurve: React.FC<Props> = ({
...
+          /**
+           * We have both the flattened and unflattened values in this TrialHParams
+           * From `const flatHParams = { ...trial.hparams, ...flattenObject(trial.hparams || {}) };`
+           * below in the file. Skip the non flattened dictionaries.
+           */
+          return;
+        }
```
[the remark this answers is not in the document]
Comment (hkang1) [reply webui/react/src/pages/ExperimentDetails/ExperimentVisualization/LearningCurve.tsx]: oh right, they are already flattened by this point.
